File: um/csharp/UniversalMachine.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Text;

namespace UniversalMachine
{
	class UM
	{
		private List<uint[]> allocatedMem = new List<uint[]> ();
		private List<int> freed = new List<int> ();

		// Registers.
		private uint[] regs = new uint[8];

		// Program counter.
		private uint pc = 0;

		// Input and output streams.
		private Stream input;
		private Stream output;

		public UM (string programPath) : this(programPath, Console.OpenStandardInput (), Console.OpenStandardOutput ())
		{
		}

		public UM (string programPath, Stream input, Stream output)
		{
			List<uint> program = new List<uint> ();
			BinaryReader reader = new BinaryReader (File.OpenRead (programPath));
			
			uint current;
			try 
			{
				while (true) 
				{
					byte b1 = reader.ReadByte ();
					byte b2 = reader.ReadByte ();
					byte b3 = reader.ReadByte ();
					byte b4 = reader.ReadByte ();
					current = (uint)((b1 << 24) | (b2 << 16) | (b3 << 8) | b4);
					program.Add (current);
				}
			} 
			catch (EndOfStreamException)
			{
			}
			
			// Add the program as the first item - index 0.
			allocatedMem.Add (program.ToArray ());
			
			this.input = input;
			this.output = output;
		}

		// This is the instruction format.
		//
		//                              A     C
		//                              |     |
		//                              vvv   vvv                    
		//      .--------------------------------.
		//      |VUTSRQPONMLKJIHGFEDCBA9876543210|
		//      `--------------------------------'
		//       ^^^^                      ^^^
		//       |                         |
		//       operator number           B
		//
		//      Figure 2. Standard Operators

		public void Run ()
		{
			bool halt = false;
			while (!halt) {
				uint instr = allocatedMem[0][pc];
				
				// Extract the register "indexes" and the instruction.
				uint a = (instr >> 6) & 7;
				uint b = (instr >> 3) & 7;
				uint c = instr & 7;
				uint opCode = instr >> 28;
				
				switch (opCode) {
				case 0:
					// Conditional move.
					// The register A receives the value in register B,
					// unless the register C contains 0.
					if (regs[c] != 0) {
						regs[a] = regs[b];
					}
					break;
				case 1:
					// Array Index.
					// The register A receives the value stored at offset
					// in register C in the array identified by B.
					regs[a] = allocatedMem[(int)regs[b]][regs[c]];
					break;
				case 2:
					// Array Amendment.
					// The array identified by A is amended at the offset
					// in register B to store the value in register C.
					allocatedMem[(int)regs[a]][regs[b]] = regs[c];
					break;
				case 3:
					// Addition.
					// The register A receives the value in register B plus 
					// the value in register C, modulo 2^32.
					regs[a] = unchecked(regs[b] + regs[c]);
					break;
				case 4:
					// Multiplication.
					// The register A receives the value in register B times
					// the value in register C, modulo 2^32.
					regs[a] = unchecked(regs[b] * regs[c]);
					break;
				case 5:
					// Division.
					// The register A receives the value in register B
					// divided by the value in register C, if any, where
					// each quantity is treated treated as an unsigned 32
					// bit number.
					regs[a] = regs[b] / regs[c];
					break;
				case 6:
					// Not-And.
					// Each bit in the register A receives the 1 bit if
					// either register B or register C has a 0 bit in that
					// position.  Otherwise the bit in register A receives
					// the 0 bit.
					regs[a] = ~(regs[b] & regs[c]);
					break;
				case 7:
					halt = true;
					break;
				case 8:
					// Allocation.
					// A new array is created with a capacity of platters
					// commensurate to the value in the register C. This
					// new array is initialized entirely with platters
					// holding the value 0. A bit pattern not consisting of
					// exclusively the 0 bit, and that identifies no other
					// active allocated array, is placed in the B register.
					uint[] newArr = new uint[regs[c]];
					// See if we have any free memory positions to reuse.
					if (freed.Count > 0) {
						// We take the index at the end of the list
						// because removing that item is much faster.
						int index = freed[freed.Count - 1];
						allocatedMem[index] = newArr;
						freed.RemoveAt (freed.Count - 1);
						regs[b] = (uint)index;
					} else {
						allocatedMem.Add (newArr);
						regs[b] = (uint)allocatedMem.Count - 1;
					}
					break;
				case 9:
					// Abandonment.
					// The array identified by the register C is abandoned.
					// Future allocations may then reuse that identifier.
					freed.Add ((int)regs[c]);
					break;
				case 10:
					// Output.
					// The value in the register C is displayed on the console
					// immediately. Only values between and including 0 and 255
					// are allowed.
					output.WriteByte ((byte)regs[c]);
					break;
				case 11:
					// Input.
					// The universal machine waits for input on the console.
					// When input arrives, the register C is loaded with the
					// input, which must be between and including 0 and 255.
					// If the end of input has been signaled, then the 
					// register C is endowed with a uniform value pattern
					// where every place is pregnant with the 1 bit.
					int i = input.ReadByte ();
					if (i == -1) {
						regs[c] = 0xFFFFFFFF;
					} else {
						regs[c] = (uint)i;
					}
					break;
				case 12:
					// Load Program.
					
					// The array identified by the B register is duplicated
					// and the duplicate shall replace the '0' array,
					// regardless of size. The execution finger is placed
					// to indicate the platter of this array that is
					// described by the offset given in C, where the value
					// 0 denotes the first platter, 1 the second, et
					// cetera.
					//
					// The '0' array shall be the most sublime choice for
					// loading, and shall be handled with the utmost
					// velocity.
					
					// If the array is the 0 array, we're just dealing with a jmp.
					if (regs[b] != 0) {
						uint[] arr = allocatedMem[(int)regs[b]];
						uint[] dup = new uint[arr.Length];
						arr.CopyTo (dup, 0);
						allocatedMem[0] = dup;
					}
					
					pc = regs[c] - 1;
					break;
				case 13:
					// Orthography.
					// The value indicated is loaded into the register A
					// forthwith.
					
					// Register A is in another place here.
					a = (instr >> 25) & 7;
					uint val = instr & 0x1FFFFFF;
					regs[a] = val;
					break;
					
				}
				
				pc++;
			}
			
		}
	}

	class MainClass
	{
		// On unix the following is an OK way of extracting UMIX.
		// echo -e "(\\\b.bb)(\\\v.vv)06FHPVboundvarHRAk\np" | mono UniversalMachine.exe codex.umz > umix
		// Then, with an editor, keep only the text that follows "UM program follows colon:".
		public static void Main (string[] args)
		{
			if (args.Length == 1) {
				UM um = new UM (args[0]);
				um.Run ();
			} else {
				//Console.WriteLine("Specify the file to load as an argument.");
				UM um = new UM ("sandmark.umz");
				um.Run ();
			}
		}
	}
}

File: um/scala/UniversalMachine.scala
import java.io.{InputStream, FileInputStream, IOException, File}
import java.util.ArrayList

/**
 * Created by IntelliJ IDEA.
 * User: simon
 * Date: 2012-01-30
 * Time: 20:11
 * Implementation of the Universal Machine.
 * Pretty much a direct translation of my Java version of this program.
 */

object UniversalMachine {
  private val allocatedMem = new ArrayList[Array[Int]]
  private val freed = new ArrayList[Int]
  private val regs = new Array[Int](8)

  private var pc: Int = 0

  def openFile(programPath: String): (InputStream, Long) = {
    val file = new File(programPath)
    if (file.length() % 4 != 0) {
      throw new IOException("File is corrupt - the number of bytes is not a multiple of 4")
    }
    (new FileInputStream(file), file.length())
  }

  def readFile(programPath: String) {
    val (inputStream, length) = openFile(programPath)
    // Read the program contained in inputStream as a sequence of (length/4) 32-bit words.
    val program = new Array[Int]((length / 4).intValue)
    val bytes = new Array[Byte](4)
    var pos = 0;
    while (inputStream.read(bytes) != -1) {
      program(pos) = ((bytes(0) & 0xFF) << 24) | ((bytes(1) & 0xFF) << 16) | ((bytes(2) & 0xFF) << 8) | (bytes(3) & 0xFF)
      pos += 1
    }
    allocatedMem.add(program)
  }
  
  def run() {
    var halt = false;
    while(!halt) {
      val instr = allocatedMem.get(0)(pc)
      val (a, b, c) = ((instr >> 6) & 7, (instr >> 3) & 7, instr & 7)
      val opCode = instr >>> 28;

      opCode match {
        case 0 => if (regs(c) != 0){ regs(a) = regs(b) }
        case 1 => regs(a) = allocatedMem.get(regs(b))(regs(c))
        case 2 => allocatedMem.get(regs(a))(regs(b)) = regs(c)
        case 3 => regs(a) = regs(b) + regs(c)
        // We want to treat the Ints as unsigned.
        case 4 => regs(a) = ((regs(b) & 0xFFFFFFFFL) * (regs(c) & 0xFFFFFFFFL)).intValue
        case 5 => regs(a) = ((regs(b) & 0xFFFFFFFFL) /  (regs(c) & 0xFFFFFFFFL)).intValue
        case 6 => regs(a) = ~(regs(b) & regs(c));
        case 7 => halt = true
        case 8 => {
          // Allocate new memory.
          val allocated = new Array[Int](regs(c))
          if (freed.size() > 0) {
            val index = freed.get(freed.size() - 1)
            allocatedMem.set(index, allocated)
            freed.remove(freed.size() - 1)
            regs(b) = index;
          } else {
            allocatedMem.add(allocated)
            regs(b) = allocatedMem.size() - 1
          }
        }
        case 9 => freed.add(regs(c))
        case 10 => printf("%c", regs(c))
        case 11 => regs(c) = readByte()
        case 12 => {
          // Load program.
          // Only if we're not dealing with the 0 array do we need to copy.
          if (regs(b) != 0) {
            val dup = allocatedMem.get(regs(b)).clone()
            allocatedMem.set(0, dup)
          }
          pc = regs(c) - 1 // We add 1 before the end of the loop.
        }
        case 13 => {
          // Orthography.
          val a = (instr >> 25) & 7;
          regs(a) = instr & 0x1FFFFFF;
        }
        case _ => {
          println("Unknown opcode encountered: " + opCode)
          halt = true;
        }
      }

      pc += 1
    }
  }

  def main(args: Array[String]) {
    readFile("sandmark.umz")
    run()
  }
}